openstack/networking/v2/subnets/requests.go

```diff
@@ -79,7 +79,7 @@ type CreateOpts struct {
 	Name            string                `json:"name,omitempty"`
 	TenantID        string                `json:"tenant_id,omitempty"`
 	AllocationPools []AllocationPool      `json:"allocation_pools,omitempty"`
-	GatewayIP       *string               `json:"gateway_ip"`
+	GatewayIP       *string               `json:"-"`
```

Comment:
OK, I've done some testing, and I think I see where the discrepancy lies:

> 2) User does not want a gateway to exist:

> map[network_id:12345 cidr:1.2.3.0/24 gateway_ip:]

Not exactly. What we want is `gateway_ip:null` (or equivalently `gateway_ip:<nil>`). What's above represents `gateway_ip:""`, which OpenStack indeed does not accept.

With that in mind, I agree that this cannot be done as-is. Adding `omitempty` to the struct tag and modifying `ToSubnetCreateMap` to the following should fix it:

```go
func (opts CreateOpts) ToSubnetCreateMap() (map[string]interface{}, error) {
	b, err := gophercloud.BuildRequestBody(opts, "subnet")
	if err != nil {
		return nil, err
	}

	if m := b["subnet"].(map[string]interface{}); m["gateway_ip"] == "" {
		m["gateway_ip"] = nil
	}

	return b, nil
}
```

With those 2 changes:
- not providing a value for `CreateOpts.GatewayIP` will give the user the OpenStack default gateway
- providing the [a pointer to] the empty string will result in the subnet not having a gateway.

Now, is that reasonable or should it be flipped, where the user has to explicitly ask for the default gateway and otherwise gets no gateway?